src/aggregator/aggregator/counter_elem_gen.go

```diff
@@ -412,28 +412,60 @@ func (e *CounterElem) processValueWithAggregationLock(
 	)
 	for aggTypeIdx, aggType := range e.aggTypes {
 		value := lockedAgg.aggregation.ValueOf(aggType)
-		for i := 0; i < transformations.Len(); i++ {
-			transformType := transformations.At(i).Transformation.Type
-			if transformType.IsUnaryTransform() {
-				fn := transformType.MustUnaryTransform()
-				res := fn(transformation.Datapoint{TimeNanos: timeNanos, Value: value})
+		for _, transformOp := range transformations {
+			unaryOp, isUnaryOp := transformOp.UnaryTransform()
+			binaryOp, isBinaryOp := transformOp.BinaryTransform()
+			switch {
+			case isUnaryOp:
+				curr := transformation.Datapoint{
+					TimeNanos: timeNanos,
+					Value:     value,
+				}
+
+				res := unaryOp.Evaluate(curr)
+
 				value = res.Value
-			} else {
-				fn := transformType.MustBinaryTransform()
-				prev := transformation.Datapoint{TimeNanos: e.lastConsumedAtNanos, Value: e.lastConsumedValues[aggTypeIdx]}
-				curr := transformation.Datapoint{TimeNanos: timeNanos, Value: value}
-				res := fn(prev, curr)
+
+			case isBinaryOp:
+				lastTimeNanos := e.lastConsumedAtNanos
+				if v := e.lastConsumedValues[aggTypeIdx].TimeNanos; v > 0 {
+					lastTimeNanos = v
+				}
+
+				prev := transformation.Datapoint{
+					TimeNanos: lastTimeNanos,
+					Value:     e.lastConsumedValues[aggTypeIdx].Value,
+				}
+
+				currTimeNanos := timeNanos
+				if t := lockedAgg.aggregation.LastAt(); !t.IsZero() {
+					currTimeNanos = t.UnixNano()
+				}
+
+				curr := transformation.Datapoint{
+					TimeNanos: currTimeNanos,
+					Value:     value,
+				}
+
+				res := binaryOp.Evaluate(prev, curr)
+
 				// NB: we only need to record the value needed for derivative transformations.
 				// We currently only support first-order derivative transformations so we only
 				// need to keep one value. In the future if we need to support higher-order
 				// derivative transformations, we need to store an array of values here.
-				e.lastConsumedValues[aggTypeIdx] = value
+				if !math.IsNaN(curr.Value) {
```

Comment:
Discussed offline why we bother passing evaluating the op if incoming sample is NaN but there may be transformations (such as count # samples) that still care.

Reply:
👍